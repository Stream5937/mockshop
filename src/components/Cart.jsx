import { Link, useOutletContext } from "react-router-dom";
import style from "../styles/cart.module.css";

export default function Cart() {
  const context = useOutletContext();
  const { products, error, loading } = context[0];
  const [cart, setCart] = context[1];
  console.log("cart: ", cart);
  const [removeFromCart, clearCart, sumCart] = context[2];
  const [quantity, setQuantity] = context[3];
  const [totalItems, setTotalItems] = context[4];

  return (
    <>
      <div className={style.cartContainer}>
        {/* Mini Cart */}

        <h2>ðŸ›’ Mini Cart (total items {totalItems})</h2>
        <div id={style.viewFull}>
          {cart.length > 0 && (
            <Link
              to="/cartPage"
              style={{ display: "block", marginTop: "0.5rem" }}
            >
              View Full Cart Page
            </Link>
          )}
        </div>

        {cart.length === 0 ? (
          <p>No items yet</p>
        ) : (
          <ul
            id={style.cart_ul}
            style={{ listStyle: "none", paddingLeft: 0, margin: 0 }}
          >
            {cart.map((obj) => (
              <li key={obj.item.id} style={{ marginBottom: "0.5rem" }}>
                <img
                  src={obj.item.image}
                  alt={obj.item.title}
                  style={{
                    verticalAlign: "middle",
                    objectFit: "contain",
                  }}
                />
                <span id={style.text} style={{ marginLeft: "0.5rem" }}>
                  {obj.item.title.substring(0, 20)}...
                  <span id={style.num}>(x{obj.number})</span>
                </span>
                <button
                  id={style.li_btn}
                  value={obj.item.id}
                  onClick={() => removeFromCart(obj.item.id)}
                  style={{ marginLeft: "0.5rem" }}
                >
                  âœ–
                </button>
              </li>
            ))}
          </ul>
        )}
      </div>
    </>
  );
}
