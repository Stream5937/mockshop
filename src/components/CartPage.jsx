import style from "../styles/cartPage.module.css";
import { useOutletContext, Link, useParams } from "react-router-dom";
import Cart from "./Cart.jsx";
import IncDecBtn from "./incDecBtn.jsx";

export default function CartPage() {
  const context = useOutletContext();
  const { products, error, loading } = context[0];
  const [cart, setCart] = context[1];
  const [removeFromCart, clearCart, sumCart] = context[2];
  const [quantity, setQuantity] = context[3];
  const [totalItems, setTotalItems] = context[4];

  return (
    <>
      <div id={style.cartContainer}>
        <div id={style.products}>
          <h2 id={style.header}>MOCK SHOP CART DETAILS</h2>
          <ul style={{ listStyle: "none", paddingLeft: 0 }}>
            {cart.map((obj) => (
              <li key={obj.item.id} style={{ marginBottom: "0.5rem" }}>
                <div id={style.itemContainer}>
                  <div id={style.product}>
                    <h3>{obj.item.title}</h3>

                    <img
                      id={style.imgDetail}
                      src={obj.item.image}
                      alt={obj.item.title}
                    />
                    <div id={style.removebtn}>
                      <button
                        value={obj.item.id}
                        onClick={() => removeFromCart(obj.item.id)}
                        style={{
                          marginTop: "0.5rem",
                          width: "200px",
                        }}
                      >
                        <span id={style.remove}>Remove item</span> ✖
                      </button>
                    </div>
                  </div>

                  <div id={style.info}>
                    <p id={style.pricecat}>
                      <span id={style.price}>
                        <b>Price: ${obj.item.price}</b>
                      </span>
                      <span id={style.category}>
                        Category: {obj.item.category}
                      </span>
                    </p>
                    <p id={style.desc}>{obj.item.description}</p>

                    <div id={style.ratecount}>
                      <p>
                        ⭐ {obj.item.rating.rate} ({obj.item.rating.count}{" "}
                        reviews)
                      </p>
                      <div id={style.incdecbtn}>
                        <IncDecBtn />
                      </div>
                    </div>
                  </div>
                </div>
              </li>
            ))}
          </ul>
        </div>
        <div id={style.cart}>
          <Cart />
        </div>
      </div>
    </>
  );
} /*}

/*
return (
    <>
      <div className={style.cartContainer}>
        <div id={style.products}>
          <h2 id={style.header}>MOCK SHOP CART DETAILS</h2>
          <ul style={{ listStyle: "none", paddingLeft: 0 }}>
            {cart.map((obj) => (
              <li key={obj.item.id} style={{ marginBottom: "0.5rem" }}>
                <div id={style.product}>
                  <h2>{obj.item.title}</h2>
                  <img
                    id={style.imgDetail}
                    src={obj.item.image}
                    alt={obj.item.title}
                  />
                </div>
                <div id={style.info}>
                  <p>
                    <b>Price: ${obj.item.price}</b>
                  </p>
                  <p>{obj.item.description}</p>
                  <p>Category: {obj.item.category}</p>
                  <p>
                    ⭐ {obj.item.rating.rate} ({obj.item.rating.count} reviews)
                  </p>
                </div>

                {/*
                <img
                  src={obj.item.image}
                  alt={obj.item.title}
                  style={{
                    width: "80px",
                    verticalAlign: "middle",
                    objectFit: "contain",
                  }}
                />
                <span style={{ marginLeft: "0.5rem" }}>
                  {obj.item.title.substring(0, 20)}... (x{obj.number})
                </span>
                */ /*
                <button
                  value={obj.item.id}
                  onClick={() => removeFromCart(obj.item.id)}
                  style={{ marginLeft: "0.5rem" }}
                >
                  ✖
                </button>
              </li>
            ))}
          </ul>
        </div>
        <div id={style.cart}>
          <Cart />
        </div>
      </div>
    </>
  );
*/
