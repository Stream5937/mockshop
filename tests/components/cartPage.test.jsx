import { describe, it, expect, vi, beforeEach } from "vitest";
import { screen, fireEvent } from "@testing-library/react";
import { renderWithRouter } from "../utils/renderWithRouter.jsx";
import CartPage from "../../src/components/CartPage.jsx";
import Cart from "../../src/components/Cart.jsx";

const mockProduct = {
  id: 1,
  title: "Test Product",
  category: "Test Category",
  price: 10,
  description: "This is a test product description",
  rating: { rate: 4.5, count: 20 },
  image: "test.jpg",
};

const mockQuantity = 3;
const mockTotalItems = 3;
const mockTotalCost = 30;

const defaultContext = [
  { products: [mockProduct], error: null, loading: false },
  [[{ item: mockProduct, number: 3 }], vi.fn()],
  [vi.fn(), vi.fn(), vi.fn()],
  [mockQuantity, vi.fn()],
  [mockTotalItems, vi.fn()],
  [mockTotalCost, vi.fn()],
];

// This variable will be used by the mock
let outletContext = defaultContext;

vi.mock("react-router-dom", async () => {
  const mod = await vi.importActual("react-router-dom");
  return {
    ...mod,
    useOutletContext: () => outletContext,
  };
});

describe("Cart Page component", () => {
  beforeEach(() => {
    outletContext = defaultContext;
  });

  it("renders", () => {
    const { asFragment } = renderWithRouter(<CartPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should find multiple h2 elements", () => {
    renderWithRouter(<CartPage />);
    const headings = screen.getAllByRole("heading", { level: 2 });
    expect(headings).toHaveLength(2);
    expect(headings[0]).toHaveTextContent("MOCK SHOP CART DETAILS");
    expect(headings[1]).toHaveTextContent("ðŸ›’ Mini Cart (total items 3)");
  });

  it("renders Remove Item, Increase and Decrease buttons", () => {
    renderWithRouter(<CartPage />);
    expect(screen.getByRole("button", { name: /Remove/i })).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /Increase/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /Decrease/i })
    ).toBeInTheDocument();
  });

  /*
  it("calls increase() when Increase Number is clicked", async () => {
    const increase = vi.fn();

    outletContext = [
      { products: [mockProduct], error: null, loading: false },
      [[{ item: mockProduct, number: 3 }], vi.fn()],
      [increase, vi.fn(), vi.fn()],
      [3, vi.fn()],
      [3, vi.fn()],
      [30, vi.fn()],
    ];

    renderWithRouter(<CartPage />);

    const increaseBtn = screen.getByRole("button", { name: /Increase/i });
    fireEvent.click(increaseBtn);

    const { default: CartDynamic } = await import(
      "../../src/components/Cart.jsx"
    );

    renderWithRouter(<CartDynamic />);
    const headings = screen.getAllByRole("heading", { level: 2 });
    // expect(headings).toHaveLength(2);
    expect(headings[0]).toHaveTextContent("MOCK SHOP CART DETAILS");
    expect(headings[1]).toHaveTextContent("ðŸ›’ Mini Cart (total items 4)");

    // expect(headings2[1]).toHaveTextContent("ðŸ›’ Mini Cart (total items 4)");
    //expect(increase).toHaveBeenCalledWith({ item: mockProduct, number: 3 });
    // Assert that the cart item quantity increased, or the UI updated
    //expect(screen.getAllByText(/4/)).toBeInTheDocument();
  });
  */
});
