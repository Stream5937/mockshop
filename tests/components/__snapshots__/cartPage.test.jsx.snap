// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Page component > renders 1`] = `
<DocumentFragment>
  <div
    id="_cartContainer_df010c"
  >
    <div
      id="_products_df010c"
    >
      <h2
        id="_header_df010c"
      >
        MOCK SHOP CART DETAILS
      </h2>
      <ul
        style="list-style: none; padding-left: 0px;"
      >
        <li
          style="margin-bottom: 0.5rem;"
        >
          <div
            id="_itemContainer_df010c"
          >
            <div
              id="_product_df010c"
            >
              <h3>
                Test Product
              </h3>
              <img
                alt="Test Product"
                id="_imgDetail_df010c"
                src="test.jpg"
              />
              <div
                id="_removebtn_df010c"
              >
                <button
                  name="remove"
                  style="margin-top: 0.5rem; width: 200px;"
                  value="1"
                >
                  <span
                    id="_remove_df010c"
                  >
                    Remove item
                  </span>
                   ‚úñ
                </button>
              </div>
            </div>
            <div
              id="_info_df010c"
            >
              <p
                id="_pricecat_df010c"
              >
                <span
                  id="_price_df010c"
                >
                  <b>
                    Price: $10
                  </b>
                </span>
                <span
                  id="_category_df010c"
                >
                  Category: Test Category
                </span>
              </p>
              <p
                id="_desc_df010c"
              >
                This is a test product description
              </p>
              <div
                id="_ratecount_df010c"
              >
                <p>
                  ‚≠ê 4.5 (20 reviews)
                </p>
                <div
                  id="_incbtn_df010c"
                >
                  <button
                    name="increase"
                    style="margin-top: 0.5rem; width: 200px;"
                    value="[object Object]"
                  >
                    <span
                      id="_inc_df010c"
                    >
                      Increase number
                    </span>
                     +
                  </button>
                </div>
                <div
                  id="_decbtn_df010c"
                >
                  <button
                    name="decrease"
                    style="margin-top: 0.5rem; width: 200px;"
                    value="[object Object]"
                  >
                    <span
                      id="_dec_df010c"
                    >
                      Decrease Number
                    </span>
                     -
                  </button>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      id="_cart_df010c"
    >
      <div
        class="_cartContainer_6ddd39"
      >
        <h2>
          üõí Mini Cart (total items 3)
        </h2>
        <div
          id="_viewFull_6ddd39"
        >
          <a
            data-discover="true"
            href="/cartPage"
            style="display: block; margin-top: 0.5rem;"
          >
            View Full Cart Page
          </a>
        </div>
        <ul
          id="_cart_ul_6ddd39"
          style="list-style: none; padding-left: 0px; margin: 0px;"
        >
          <li
            style="margin-bottom: 0.5rem;"
          >
            <img
              alt="Test Product"
              src="test.jpg"
              style="vertical-align: middle; object-fit: contain;"
            />
            <span
              id="_text_6ddd39"
              style="margin-left: 0.5rem;"
            >
              Test Product...
              <span
                id="_num_6ddd39"
              >
                (x3)
              </span>
            </span>
            <button
              id="_li_btn_6ddd39"
              style="margin-left: 0.5rem;"
              value="1"
            >
              ‚úñ
            </button>
          </li>
        </ul>
        <div
          id="_cartTotal_6ddd39"
        >
          <p>
            Current Total to Pay: $30 
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
